heap: some cleanup and make configurable

nashif opened the pull request:
We have been building heap related code unconditionally for everyone and heap implementation and tests were spread across the tree. Moving everything under lib/heap and splitting heap-validate into smaller manageable and configurable chunks and consolidate tests under test/lib.




- lib: os: move heap to own folder
- lib: heap: split heap-validate.c
- lib: heap: build heap utilities only on demand
- lib: heap: make heap_print_info static
- tests: move mheap_api_concept into tests/lib/multi_heap
- tests: multi_heap: adap test identifiers
- tests: move shared_multi_heap to tests/lib
- tests: shared_multi_heap: fix identifier
- lib: heap: rename with files and use _
- lib: heap: build heap_info conditionally
- lib: heap: make multi-heap feature configurable

nashif committed:
lib: os: move heap to own folder

consolidate all heap related files and implementation into one folder.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
lib: heap: split heap-validate.c

Split heap-validate.c into smaller chunks. We have been adding all kind
of new APIs under this file and building it unconditionally whether
those APIs are needed/used or not. Many of those APIs have nothing to do
with the validation part.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
lib: heap: build heap utilities only on demand

Build the heap features only when needed and based on Kconfigs being
set.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
lib: heap: make heap_print_info static

This should be declated static as it is being only called in this file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: move mheap_api_concept into tests/lib/multi_heap

Move multi-heap tests out of kernel folder along side all other heap
related tests.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: multi_heap: adapt test identifiers

Change identifier to match with other heap related tests.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: move shared_multi_heap to tests/lib

Move shared_multi_heap test under lib alongside all other heap tests.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: shared_multi_heap: fix identifier

Adapt identifier and be consistent with other heap tests.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
lib: heap: rename with files and use _

be consistent with other files in the same folder.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
lib: heap: build heap_info conditionally

Only build when needed and adapt existing users to pull this in when
needed.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
lib: heap: make multi-heap feature configurable

Make multi-heap configurable via Kconfig and adapt existing tests.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

npitre commented:
There are options in lib/heap/Kconfig that don't belong there:

`SYS_MEM_BLOCKS`
`SYS_MEM_BLOCKS_LISTENER`
`SYS_MEM_BLOCKS_RUNTIME_STATS`
`OBJ_CORE_SYS_MEM_BLOCKS`
`OBJ_CORE_STATS_SYS_MEM_BLOCKS`

Those refer to lib/os/mem_blocks.c which is independent from the heap code.
I'd suggest creating lib/mem_blocks/, moving lib/os/mem_blocks.c there and
splitting it similar to what this PR does with the heap code.

nashif committed:
lib: mem_block: move to own folder

Move mem_block into own folder and seperate from lib/os and heap
configuration.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

npitre commented:
Anas: you forgot to remove `SYS_MEM_BLOCKS` and friends from heap/Kconfig.

nashif committed:
lib: heap: move heap stats to own file

heap stats are split out from heap_validate.c into own file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
preexisting checkpatch issues:

```
 UNNECESSARY_ELSE: else is not generally useful after a break or return
File:lib/heap/heap_stress.c
Line:48
 MACRO_WITH_FLOW_CONTROL: Macros with flow control statements should be avoided
File:lib/heap/heap_validate.c
Line:21
```


@andyross  @npitre  please re-review
